File: day03/main.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
)

const (
	puzzleInput = "input.txt"
)

func main() {
	file, err := os.Open(puzzleInput)
	if err != nil {
		log.Fatal(err)
	}
	defer func() {
		if err := file.Close(); err != nil {
			log.Fatal(err)
		}
	}()

	m := [][]byte{}

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		m = append(m, []byte(scanner.Text()))
	}
	if err := scanner.Err(); err != nil {
		log.Fatal(err)
	}

	fmt.Println(getCount(3, 1, m))
	fmt.Println(getCount(1, 1, m) * getCount(3, 1, m) * getCount(5, 1, m) * getCount(7, 1, m) * getCount(1, 2, m))
}

func getCount(x, y int, m [][]byte) int {
	count := 0
	i := 0
	j := 0
	for i < len(m) {
		if m[i][j] == '#' {
			count++
		}
		i += y
		j = (j + x) % len(m[0])
	}
	return count
}

File: day03/src/main.rs
use std::fs::File;
use std::io::prelude::*;
use std::io::BufReader;

const PUZZLEINPUT: &str = "input.txt";

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let file = File::open(PUZZLEINPUT)?;
    let reader = BufReader::new(file);

    let m = {
        let mut m = Vec::new();
        for line in reader.lines() {
            m.push(line?.chars().collect::<Vec<char>>());
        }
        m
    };

    println!("Part 1: {}", get_count(3, 1, &m));
    println!(
        "Part 2: {}",
        get_count(1, 1, &m)
            * get_count(3, 1, &m)
            * get_count(5, 1, &m)
            * get_count(7, 1, &m)
            * get_count(1, 2, &m)
    );

    Ok(())
}

fn get_count(x: usize, y: usize, m: &Vec<Vec<char>>) -> usize {
    let mut count = 0;
    let mut i = 0;
    let mut j = 0;
    let w = m[0].len();
    while i < m.len() {
        if m[i][j] == '#' {
            count += 1;
        }
        i += y;
        j = (j + x) % w;
    }
    count
}
